File: kinosan.php
<?php
//$time_start_ex = microtime(true);
define('SL',31);	//分割する桁数（とりあえず32ビットの正の数分=31桁）
stream_set_blocking(STDIN, 0);
$stdin = file_get_contents('php://stdin');
$first = strpos($stdin,"\n");
list($H,$W)=explode(' ',substr($stdin,0,$first));
$stdin = substr($stdin,$first+1);
$lines = explode("\n", $stdin);
$home_lines = array_slice($lines, 0, $H);
$widgets = array_slice($lines, $H + 1, $lines[$H]);
//$time_end = microtime(true);
//$time = ($time_end - $time_start_ex) * 1000;
//echo "!!!INPUT_TIME->$time!!!".PHP_EOL;
//print_r('!!!check!!!');
//print_r($home_lines);
//print_r($widgets);

if ($W > SL){
	//ホームの幅が広い場合はSL幅で計算
	$WW = SL;
}
else {
	$WW = $W;
}

// $bin_max_val = pow(2,$WW) - 1;		//2進数最大値	
// $str_max_val = str_repeat('1',$WW);	//文字列最大値
//$str_min_val = str_repeat('0',$WW);		//文字列最小値

//衝突チェックブロック配列
$check_block = array();
//スキマチェック用リスト
$check_list = array();
for ($SH=1;$SH<=$H;++$SH){
//	$check_block[$SH] = array();
	$check_list[$SH] = array();
}

//$SORT_COND = $H - 1;
foreach ($home_lines as $j => $hl_val){
	//echo "-----j->$j\n";
	//[最初の準備] ホームの情報をSLの幅で分けて2進数にし、配列に格納
	$check_block[$j] = array();
	$pos = 0;
	while ($pos<$W){
		$work_home = substr($hl_val,$pos,SL);
		if (strlen($work_home)<$WW){
			//最後のブロックの右側を1で埋める
			$work_home = str_replace(array("\r\n","\r","\n"), '', $work_home);
			$work_home = str_pad($work_home,$WW,'1');
		}
		$check_block[$j][] = intval($work_home, 2);		//2進数に変換して格納
		$pos += SL;
	}

	$str_check_list = '';
	foreach ($check_block[$j] as $i => $val){
	 	$str_bin = decbin($val);
	 	if (strlen($str_bin)<$WW){
	 		$str_bin = str_pad($str_bin, $WW, '0', STR_PAD_LEFT);
	 	}
	 	$str_check_list .= $str_bin;
	}

	$pos = 0;
	while ($pos<$W){
		$first = strpos($str_check_list,'0',$pos);
		if ($first !== false){
			$second = strpos($str_check_list,'1',$first);
			if ($second !== false){
				$key = $second - $first;
				$pos = $second;
			}
			else {
				$key = $W - $first;
				$pos = $W;
			}
			if (!isset($check_list[1][$key])){
				$check_list[1][$key]=1;
			}
			else {
				++$check_list[1][$key];
			}
		}
		else {
			break;
		}
	}
}
if (count($check_list[1])>1){
	krsort($check_list[1]);
}

for ($SH=2;$SH<=$H;++$SH){
	$k = $H - $SH;
	$cl = &$check_list[$SH];
	for ($j=0;$j<=$k;++$j){
		$str_check_list = '';
		$cb = &$check_block[$j];
		$cbp = &$check_block[$j+1];
		foreach ($cb as $i => $val){
			//OR演算
			$cb[$i] = $cbp[$i] | $val;
			// if ($check_block[$j][$i]===0){
			// 	$str_check_list .= $str_min_val;
			// }
			// else {
			 	$str_bin = decbin($cb[$i]);
			 	if (strlen($str_bin)<$WW){
			 		$str_bin = str_pad($str_bin, $WW, '0', STR_PAD_LEFT);
			 	}
			 	$str_check_list .= $str_bin;
			// }
		}

		$pos = 0;
		while ($pos<$W){
			$first = strpos($str_check_list,'0',$pos);
			if ($first !== false){
				$second = strpos($str_check_list,'1',$first);
				if ($second !== false){
					$key = $second - $first;
					$pos = $second;
				}
				else {
					$key = $W - $first;
					$pos = $W;
				}
				if (!isset($cl[$key])){
					$cl[$key]=1;
				}
				else {
					++$cl[$key];
				}
				//print_r($check_list[$SH]);
			}
			else {
				break;
			}
		}
	}
	if (count($cl)>1){
		krsort($cl);
	}
}

//print_r($check_list);
//$time_end = microtime(true);
//$time = ($time_end - $time_start_ex) * 1000;
//echo "!!!PREPARED_TIME->$time!!!".PHP_EOL;
$echo_str = '';
foreach ($widgets as $w_val){
	//ウィジェットループ
	list($SH,$TW) = explode(' ',$w_val);
	$cnt = 0;
	if ($SH>$H && $TW>$W){
		//ウィジェットのほうが大きい
	}
	else {
		foreach ($check_list[$SH] as $key => $val){
			if ($key<$TW) break;
			$cnt += ($key-$TW+1) * $val;
		}
	}
	//$echo_str .= "SH->$SH TW->$TW ".$cnt.PHP_EOL;
	$echo_str .= $cnt.PHP_EOL;
}
echo $echo_str;
//$time_end = microtime(true);
//$time = ($time_end - $time_start_ex) * 1000;
//echo "!!!TOTAL_TIME->$time!!!".PHP_EOL;
?>

File: kinosan.pl
# use Data::Dumper;
# use Time::HiRes;  
# my $start_time = Time::HiRes::time;  
my $SL = 31;	#分割する桁数（とりあえず32ビットの正の数分=31桁）

<> =~ /^(\d+) (\d+)$/;
my $H = $1;
my $W = $2;
#print "H\n";
#print "$H\n";
#print "W\n";
#print "$W\n";

my @home_lines = map { scalar(<>) } (1..$H);
#@home_lines = map {$_ =~ s/\n//; $_} @home_lines;
map{scalar(<>)}(0);
#my @widgets = map { scalar(<>) } (0..$N-1);
#@widgets = map {$_ =~ s/\n//; $_} @widgets;

# printf("!!!INPUT_TIME->%0.3f\n",Time::HiRes::time - $start_time);

#print "home_lines\n";
#print @home_lines;

my $WW = $W;
if ($W > $SL){
	$WW = $SL;
}

my $bin_max_val = 2 ** $WW - 1;	#2進数最大値
my $str_max_val = '1' x $WW;		#文字列最大値

#print "$bin_max_val\n";
#print "$str_max_val\n";

#衝突チェックブロック配列
my @check_block;

#スキマチェック用ハッシュ
my %check_list;
keys(%check_list) = $H;

my $format1 = '%-0' . $WW . 's';
my $format2 = '%0' . $WW . 'b';
my $pos = 0;
my @bin_home_list;
my $str_check_list;
my $chk;
my $SH;
my $TW;

#my $COUNT_MAX_SH = 0;
my $NEXT_MAX_SH = $H;

foreach (0..$#home_lines) {
	#printf '[%d] %s' . "\n", $_, $home_lines[$_];

	#[最初の準備] ホームの情報をSLの幅で分けて2進数にし、配列に格納
	@bin_home_list = ();
	$pos = 0;
	while ($pos < $W) {
		my $work_home = substr($home_lines[$_], $pos, $SL);
		if (length($work_home)<$WW){
			#最後のブロックの右側を1で埋める
			$work_home = sprintf($format1, $work_home);
			$work_home=~ s/\s/1/g;
			#print "$work_home|\n";
		}
		#push @bin_home_list, oct("0b$work_home");
		$bin_home_list[@bin_home_list] = oct("0b$work_home");
		$pos += $SL;
	}

	#print Dumper @bin_home_list;

	$SH = $_ + 1;
	while ($SH > 0){
		if ($SH<=$NEXT_MAX_SH){
			# [手順1] ウィジェットの高さパターンに応じてOR演算をしたリストを作成
			#(例)
			#     00100000
			#     01000011　⇒ 01100111
			#     01100100
			$str_check_list = '';
			#$ck_block_ref = \%{$check_list{$SH-1}};
			foreach (0..$#bin_home_list) {
				if ($SH == 1){
					$chk = $bin_home_list[$_];
				}
				else {
					#OR演算
					if ($check_block{$SH-1}->{$_}==$bin_max_val) {
						$chk = $bin_max_val;
					}
					else {
						$chk = $check_block{$SH-1}->{$_} | $bin_home_list[$_];
					}
				}
				$check_block{$SH}{$_} = $chk;

				#求めた2進数を文字列に変換
				if ($chk == $bin_max_val){
					$str_check_list .= $str_max_val;
				}
				else {
					$str_check_list .= sprintf($format2,$chk);
				}

				#printf '[%d] %d %s' . "\n", $_, $SH, $str_check_list;
			}


			# [手順2] 求めた2進数からゼロ文字列の長さリストを作成
			#(例)
			#     0011000110101011　⇒　2,3,1,1,1
			$pos = 0;
			my $key;
			while ($pos < $W){
				my $first = index($str_check_list,'0',$pos);
				if ($first >= 0){
					my $second = index($str_check_list,'1',$first);
					if ($second >= 0){
						$key = $second - $first;
						$pos = $second;
					}
					else {
						$key = $W - $first;
						$pos = $W;
					}
					#if (exists($check_list{$SH}{$key})){
					 	$check_list{$SH}{$key}+=1;
					#}
					#else {
					#	$check_list{$SH}{$key}=1;
					#}
				}
				else {
					if ($pos==0){
						#ゼロが存在しない　→　次のループより入りきるウィジェット高さが制限される
						$NEXT_MAX_SH = $SH;
					}
					last;
				}
			}
			$SH -= 1;
		}
		else {
			$SH = $NEXT_MAX_SH;
		}
	}
	#print "-----block\n";
	#print Dumper @check_block;
}

#print "-----check_list\n";
#print Dumper %check_list;

# printf("!!!PREPARED_TIME->%0.3f\n",Time::HiRes::time - $start_time);

my $print_str = '';
my $cnt;
for my $w_val (<>) {
#foreach my $w_val (@widgets){
	#ウィジェットループ
	($SH,$TW) = split(/ /,$w_val);
	#$TW = chomp($TW);
	#print "SH->$SH\n";
	#print "TW->$TW\n";
	$cnt = 0;
	if ($SH<=$H){
		foreach (sort {$b <=> $a} (keys %{$check_list{$SH}}))  {
		#foreach (keys %{$check_list{$SH}})  {
			#printf '%d [%d] %d' . "\n", $SH, $_, $check_list{$SH}->{$_};
			if ($_ < $TW){
				last;
			}
			$cnt += ($_-$TW+1) * $check_list{$SH}->{$_};
		}
	}
	$print_str .= $cnt."\n";
}
print $print_str;
# printf("!!!TOTAL_TIME->%0.3f\n",Time::HiRes::time - $start_time);
